Test VRF v2 locally using a mock contract (#1219)
smartcontractkit/documentation: src/pages/vrf/v2/subscription/examples/test-locally.mdx

---
layout: ../../../../../layouts/MainLayout.astro
section: vrf
title: "Local testing using a Mock contract"
metadata:
  description: "Example contract for generating random words using the VRF v2 subscription method on your local blockchain using a mock contract."
---

import VrfCommon from "@features/vrf/v2/common/VrfCommon.astro"
import ContentCommon from "@features/common/ContentCommon.astro"
import { CodeSample, ClickToZoom, Aside } from "@components"

<VrfCommon callout="subscription" />

> This guide explains how to test Chainlink VRF v2 on a [Remix IDE](https://remix-ide.readthedocs.io/en/latest/run.html#environment) sandbox blockchain environment. **Note**: You can reuse the same logic on another development environment, such as hardhat or truffle. For example, read the hardhat starter kit [RandomNumberConsumer unit tests](https://github.com/smartcontractkit/hardhat-starter-kit/blob/main/test/unit/RandomNumberConsumer.spec.js).

Reviewer: Capitalize these terms when referring to them in plain text (not code references): 

- "Hardhat"
- "Truffle"
- "Hardhat Starter Kit"

Apply throughout this document and the direct funding version.

## Testing

### Open the contracts on RemixIDE

Open _VRFCoordinatorV2Mock_ and compile in Remix:

<CodeSample src="samples/VRF/mock/VRFCoordinatorV2Mock.sol" />

Open _VRFv2Consumer_ and compile in Remix:

<CodeSample src="samples/VRF/mock/VRFv2Consumer.sol" />

Your RemixIDE file explorer should display _VRFCoordinatorV2Mock.sol_ and _VRFv2Consumer.sol_:

<ClickToZoom src="/images/vrf/mock/v2-subscription-remix-fileexplorer.jpg" />

### Deploy VRFCoordinatorV2Mock

1. Open _VRFCoordinatorV2Mock.sol_.

1. Under _DEPLOY & RUN TRANSACTIONS_, select _VRFCoordinatorV2Mock_.

   <ClickToZoom src="/images/vrf/mock/v2-deploymock.jpg" />

1. Under _DEPLOY_, fill in the `_BASEFEE` and `_GASPRICELINK`. These variables are used in the _VRFCoordinatorV2Mock_ contract to represent the base fee and the gas price (in LINK tokens) for the VRF requests. You can set: `_BASEFEE=100000000000000000` and `_GASPRICELINK=1000000000`.

1. Click on _transact_ to deploy the _VRFCoordinatorV2Mock_ contract.

> 1. Once deployed, you should see the _VRFCoordinatorV2Mock_ contract under _Deployed Contacts_.

Reviewer: ```suggestion
1. Once deployed, you should see the _VRFCoordinatorV2Mock_ contract under _Deployed Contracts_.
```

Reviewer: Search for additional instances in this file and also apply changes to direct funding file

1. Note the address of the deployed contract.

### Create and fund a subscription

1. Click on `createSubscription` to create a new subscription.

1. In the RemixIDE console, read your transaction decoded output to find the subscription ID. In this example, the subscription ID is _1_.

   <ClickToZoom src="/images/vrf/mock/v2-mocksubscription.jpg" />

1. Click on `fundSubscription` to fund your subscription. In this example, you can set the `_subid` to `1` (which is your newly created subscription ID) and the `_amount` to `1000000000000000000`.

### Deploy the VRF consumer contract

1. In the file explorer, open _VRFv2Consumer.sol_.

1. Under _DEPLOY & RUN TRANSACTIONS_, select _RandomNumberConsumerV2_.

   <ClickToZoom src="/images/vrf/mock/v2-deployconsumer.jpg" />

1. Under _DEPLOY_, fill in `SUBSCRIPTIONID` with your subscription ID, `vrfCoordinator` with the deployed _VRFCoordinatorV2Mock_ address and, _KEYHASH_ with an arbitrary `bytes32` (In this example, you can set the _KEYHASH_ to `0xd89b2bf150e3b9e13446986e571fb9cab24b13cea0a43ea20a6049a85cc807cc`).

1. Click on _transact_ to deploy the _RandomNumberConsumerV2_ contract.

> 1. Once deployed, you should see the _RandomNumberConsumerV2_ contract under _Deployed Contracts_.

Reviewer 2: ```suggestion
1. After the consumer contract is deployed, you should see the _RandomNumberConsumerV2_ contract under _Deployed Contracts_.
```

1. Click on `requestRandomWords`.

   <ClickToZoom src="/images/vrf/mock/v2-requestrandomwords.jpg" />

1. In the RemixIDE console, read your transaction logs to find the VRF request ID. In this example, the request ID is _1_.

   <ClickToZoom src="/images/vrf/mock/v2-requestrandomwordslogs.jpg" />

1. Note your request ID.

### Fulfill the VRF request

Because you are testing on a local blockchain environment, you must fulfill the VRF request yourself.

1. Under _Deployed Contracts_, open the functions list of your deployed _VRFCoordinatorV2Mock_ contract.

1. Click `fulfillRandomWords` and fill in `_requestId` with your VRF request ID and `_consumer` with your consumer contract address.

   <ClickToZoom src="/images/vrf/mock/v2-fulfillrandomwords.jpg" />

1. Click on `transact`.

### Check the results

1. Under _Deployed Contracts_, open the functions list of your deployed _RandomNumberConsumerV2_ contract.

1. Click on `s_requestId` to display the last request ID. In this example, the output is _1_.

   <ClickToZoom src="/images/vrf/mock/v2-deployedconsumerrequestid.jpg" />

> 1. Each time you make a VRF request, your consumer contract requests two random words. Once the request is fulfilled, the two random words are stored in the `s_randomWords` array. You can check the stored random words by reading the two first indexes of the `s_randomWords` array. To do so, click on the _s_randomWords_ function and:

Reviewer 2: ```suggestion
1. Each time you make a VRF request, your consumer contract requests two random words. After the request is fulfilled, the two random words are stored in the `s_randomWords` array. You can check the stored random words by reading the two first indexes of the `s_randomWords` array. To do so, click on the _s_randomWords_ function and:
```

   1. Fill in the index with _0_ then click on _call_ to read the first random word.

      {" "}

      <ClickToZoom src="/images/vrf/mock/v2-firstrandomword.jpg" />

   1. Fill in the index with _1_ then click on _call_ to read the second random word.

      {" "}

      <ClickToZoom src="/images/vrf/mock/v2-secondrandomword.jpg" />

> ## Conclusion

Reviewer: Nit: I haven't seen most docs tutorials end with a section actually titled "Conclusion" - my suggestion is to retitle this "Next steps" if it doesn't fit well into the usual structure for our "What's next" sections.

After making that change, we would avoid adding a "What's next" section to these specific pages to avoid duplication.
